src/Halogen/Query/EventSource.purs

```diff
@@ -1,186 +1,172 @@
-unEventSource (EventSource e) = e
+derive instance newtypeEventSource :: Newtype (EventSource m f) _
 
-interpret :: forall f g m. Functor m => (f ~> g) -> EventSource f m -> EventSource g m
-interpret nat (EventSource es) =
+-- | Constructs an event source from a setup function that operates in `Aff`.
+-- |
+-- | - The `Emitter` that the passed function receives is used to `emit` queries
+-- |   that will be received by the current component, or can be `close`d to
+-- |   shut down the event source and remove the subscription.
+-- | - The `Finalizer` that the passed function produces is there to allow for
+-- |   some clean-up action to be taken when the event source is unsubscribed
+-- |   from. This also runs if the `Emitter` is `close`d. `mempty` can be used
+-- |   here if there is no clean-up to perform.
+affEventSource
+  :: forall m f
+   . MonadAff m
+  => (Emitter Aff f -> Aff (Finalizer Aff))
+  -> EventSource m f
+affEventSource recv = EventSource $ liftAff do
+  inputVar <- AV.empty
+  finalizeVar <- AV.empty
+  let
+    producer = do
+      lift $ liftAff $ flip AV.put finalizeVar =<< recv (Emitter (flip AV.put inputVar))
+      CR.producer $ liftAff $ AV.take inputVar
+    finalizer = Finalizer do
+      liftAff (attempt (AV.take finalizeVar)) >>= case _ of
+        Left _ -> pure unit
+        Right z -> liftAff do
+          AV.kill (Exn.error "finalized") finalizeVar
+          finalize z
+  pure { producer, finalizer }
+
+-- | Constructs an event source from a setup function that operates in `Eff`.
+-- |
+-- | - The `Emitter` that the passed function receives is used to `emit` queries
+-- |   that will be received by the current component, or can be `close`d to
+-- |   shut down the event source and remove the subscription.
+-- | - The `Finalizer` that the passed function produces is there to allow for
+-- |   some clean-up action to be taken when the event source is unsubscribed
+-- |   from. This also runs if the `Emitter` is `close`d. `mempty` can be used
+-- |   here if there is no clean-up to perform.
+effEventSource
+  :: forall m f
+   . MonadAff m
+  => (Emitter Effect f -> Effect (Finalizer Effect))
+  -> EventSource m f
+effEventSource =
+  affEventSource <<<
+    dimap
+      (hoistEmitter launchAff_)
+      (liftEffect <<< map (hoistFinalizer liftEffect))
+
+-- | Constructs an event source from an event in the DOM. Accepts a function
+-- | that maps event values to a `Maybe`-wrapped query, allowing it to filter
+-- | events if necessary.
+eventListenerEventSource
+  :: forall m f
+   . MonadAff m
+  => E.EventType
+  -> ET.EventTarget
+  -> (E.Event -> Maybe (f Unit))
+  -> EventSource m f
+eventListenerEventSource eventType target f = effEventSource \emitter -> do
+  listener <- ET.eventListener (maybe (pure unit) (emit emitter <<< pure) <<< f)
+  ET.addEventListener eventType listener false target
```

Comment:
Have you considered to also take `capture :: Boolean` as argument?

Reply:
Not really - I just included this as a convenience for the common case. It's easy to implement with `effEventSource` anyway, so I figured there's that if people require it.

Reply:
Can you think of a case where using `capture` is common? I'd be tempted to add it as a separate named function if we could think of a situation in which listening in the capture phase was the norm.

Reply:
It's useful when you are listening to some event on body for example and might want to cancel propagation of event, or just want to change/set something before handler of some element deep in the dom is fired. but usually it's not that common.

More real example would be if you are embedding some lib in the dom and the lib on clickdoes something but also cancels event propagation. if you want to know if click happened, you could set capture to true on wrapper node and your listener will be fired before the lib get's it